sagemaker-fundamentals/create-training-job/container_kms/train.py

```diff
@@ -0,0 +1,180 @@
+
+
+# where SageMaker injects training info inside container
+input_dir="/opt/ml/input/"
+
+# where SageMaker injects train / validation /test data
+data_dir = '/opt/ml/input/data/'
+
+# SageMaker treat "/opt/ml/model" as checkpoint direcotry
+# and it will send everything there to S3 output path you 
+# specified 
+model_dir="/opt/ml/model"
+
+
+def encrypt(data:bytes, plaintext_key:bytes) -> bytes:
+    """Encrypt a chunk of bytes on client-side
+    
+    Args:
+        data: a chunk of bytes
+        plaintext_key: plaintext data key
+        
+    Return:
+        encrypted data
+    """
+    ascii_str = base64.b64encode(plaintext_key)
+    f = Fernet(key=ascii_str) 
+    return f.encrypt(data)
+
+
+def decrypt(data:bytes, plaintext_key:bytes) -> bytes:
+    """Decrypt a chunk of bytes on client-side
+    
+    Args:
+        data: encypted binary data
+        plaintext_key: plaintext data key
+    
+    Return:
+        decrypted data
+    """    
+    # to Fernet-friendly key
+    ascii_str = base64.b64encode(plaintext_key)
+    
+    f = Fernet(key=ascii_str)
+    return f.decrypt(data)
+
+    
+def get_data_key_ciphertext(bucket:str, key:str) -> bytes:
+    """Get ciphertext data key from an S3 bucket
+    
+    Args:
+        bucket: name of the S3 bucket
+        key: object key in the S3 bucket (not to be confused with encryption key)
+        
+    Return: 
+        data key ciphertext
+    """
+    obj = s3.get_object(Bucket=bucket, Key=key)
+    return obj['Body'].read()
+
+def get_data_key_plaintext(master_key:str, ciphertext: bytes) -> bytes:
+    """Decrypt ciphertext data key
+    
+    Args:
+        master_key: id of the master key
+        ciphertext: ciphertext of the data key
+    
+    Return:
+        plaintext data key
+    """
+    
+    plaintext = kms.decrypt(
+        KeyId=master_key, 
+        CiphertextBlob=ciphertext)['Plaintext']
+    return plaintext
+    
+    
+def main():
+    print("== Loading hyperparamters ===")
+    with open(
+        os.path.join(input_dir, 'config/hyperparameters.json'), 'rb') as f:
+        hyp = json.load(f)
+    
+    
+    # get ciphertext data key
+    C = get_data_key_ciphertext(
+        bucket=hyp['key_bucket'], 
+        key=hyp['ciphertext_s3_key']
+    )
+    
+    # get plaintext data key
+    P = get_data_key_plaintext(
+        master_key=hyp['master_key_id'],
+        ciphertext=C
+    )
+    
+    # read encrypted binary data
+    data_file = os.path.join(
+        data_dir, hyp['train_channel'], hyp['train_file'])
+    
+    with open(data_file, 'rb') as f:
+        data = f.read()
+        
+    # if you do not decrypt the data 
+    # then python cannot deserialize it to an object
+    try:
+        print("=== deserializing encrypted data (going to fail) ===")
+        pickle.loads(data)
+    except Exception as e:
+        print('Failed to unpickle')
+        print(e)
+    
+    # decrypt your training data
+    plain_data = decrypt(data=data, plaintext_key=P)
```

Comment:
You have no encryption context (or AAD here) ideally there should be some bounding on this data. To make sure that you have what you expect. Just because it decrypts does not mean it is the right data.

Reply:
that's true, but how do you know the encryption context is not modified?

Reply:
This is the nature of authenticated encryption. If any of the it is modified the decryption will fail, again, this is why doing crypto at this level is a bad idea.